File: app/src/main/java/com/example/dreamloaf/production/ProductionLogFragment.java
package com.example.dreamloaf.production;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import com.example.dreamloaf.adapter.ProductionAdapter;
import com.example.dreamloaf.databinding.FragmentProductionLogBinding;
import com.example.dreamloaf.viewmodels.ProductViewModel;
import com.example.dreamloaf.viewmodels.ProductionViewModel;
import com.example.dreamloaf.viewmodels.SaleViewModel;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;

public class ProductionLogFragment extends Fragment {

    private FragmentProductionLogBinding binding;
    private ProductViewModel productViewModel;
    private ProductionViewModel productionViewModel;
    private SaleViewModel saleViewModel;
    private ProductionAdapter adapter;
    private String selectedDate;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentProductionLogBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        selectedDate = getArguments() != null ?
                getArguments().getString("selected_date", getCurrentDate()) :
                getCurrentDate();

        productViewModel = new ViewModelProvider(requireActivity()).get(ProductViewModel.class);
        productionViewModel = new ViewModelProvider(this).get(ProductionViewModel.class);
        saleViewModel = new ViewModelProvider(this).get(SaleViewModel.class);
        adapter = new ProductionAdapter(productionViewModel, saleViewModel, selectedDate);
        adapter.setEditMode(true);
        binding.productionRecyclerView.setAdapter(adapter);
        binding.productionRecyclerView.setLayoutManager(new LinearLayoutManager(requireContext()));

        binding.tvDateTitle.setText("Производство на " + selectedDate);

        binding.btnSaveAll.setOnClickListener(v -> {
            adapter.saveAllData();
            Toast.makeText(requireContext(), "Данные сохранены", Toast.LENGTH_SHORT).show();
        });

        loadDataForDate(selectedDate);
    }

    private String getCurrentDate() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            return LocalDate.now().format(DateTimeFormatter.ISO_LOCAL_DATE);
        } else {
            return java.text.DateFormat.getDateInstance().format(new java.util.Date());
        }
    }

    private void loadDataForDate(String date) {
        productViewModel.getAllProducts().observe(getViewLifecycleOwner(), products -> {
            if (products.isEmpty()) {
                binding.emptyState.setVisibility(View.VISIBLE);
                binding.productionRecyclerView.setVisibility(View.GONE);
            } else {
                binding.emptyState.setVisibility(View.GONE);
                binding.productionRecyclerView.setVisibility(View.VISIBLE);

                productionViewModel.getProductionByDate(date).observe(getViewLifecycleOwner(), productions -> {
                    saleViewModel.getSalesByDate(date).observe(getViewLifecycleOwner(), sales -> {
                        adapter.updateProductionsAndSales(productions, sales);
                        adapter.submitList(products);
                    });
                });
            }
        });
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/main/java/com/example/dreamloaf/production/ProductionLogFragment.kt
package com.example.dreamloaf.production

import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.dreamloaf.adapter.ProductionAdapter
import com.example.dreamloaf.databinding.FragmentProductionLogBinding
import com.example.dreamloaf.viewmodels.ProductViewModel
import com.example.dreamloaf.viewmodels.ProductionViewModel
import com.example.dreamloaf.viewmodels.SaleViewModel
import java.text.DateFormat
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.Date

class ProductionLogFragment : androidx.fragment.app.Fragment() {
    private var binding: FragmentProductionLogBinding? = null
    private lateinit var productViewModel: ProductViewModel
    private lateinit var productionViewModel: ProductionViewModel
    private lateinit var saleViewModel: SaleViewModel
    private var adapter: ProductionAdapter? = null
    private var selectedDate: String? = null

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentProductionLogBinding.inflate(inflater, container, false)
        return binding!!.getRoot()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        selectedDate = arguments?.getString("selected_date", this.currentDate) ?: this.currentDate
        productViewModel = ViewModelProvider(requireActivity()).get(ProductViewModel::class.java)
        productionViewModel = ViewModelProvider(this).get(ProductionViewModel::class.java)
        saleViewModel = ViewModelProvider(this).get(SaleViewModel::class.java)
        adapter = ProductionAdapter(productionViewModel, saleViewModel, selectedDate)
        adapter!!.setEditMode(true)
        binding!!.productionRecyclerView.adapter = adapter
        binding!!.productionRecyclerView.layoutManager = LinearLayoutManager(requireContext())
        binding!!.tvDateTitle.text = "Производство на $selectedDate"
        binding!!.btnSaveAll.setOnClickListener {
            adapter!!.saveAllData()
            Toast.makeText(requireContext(), "Данные сохранены", Toast.LENGTH_SHORT).show()
        }
        loadDataForDate(selectedDate)
    }

    private val currentDate: String?
        get() {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                return LocalDate.now()
                    .format(DateTimeFormatter.ISO_LOCAL_DATE)
            } else {
                return DateFormat.getDateInstance().format(Date())
            }
        }

    private fun loadDataForDate(date: String?) {
        productViewModel.allProducts?.observe(viewLifecycleOwner) { products ->
            if (products.isNullOrEmpty()) {
                binding!!.emptyState.visibility = View.VISIBLE
                binding!!.productionRecyclerView.visibility = View.GONE
            } else {
                binding!!.emptyState.visibility = View.GONE
                binding!!.productionRecyclerView.visibility = View.VISIBLE
                productionViewModel.getProductionByDate(date)?.observe(viewLifecycleOwner) { productions ->
                    saleViewModel.getSalesByDate(date)?.observe(viewLifecycleOwner) { sales ->
                        val safeProductions = productions?.filterNotNull() ?: emptyList()
                        val safeSales = sales?.filterNotNull() ?: emptyList()
                        val safeProducts = products.filterNotNull()
                        adapter!!.updateProductionsAndSales(safeProductions, safeSales)
                        adapter!!.submitList(safeProducts)
                    }
                }
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }
}